File: platforms/ios/ZylixUITests/ZylixTestContext.swift
//
//  ZylixTestContext.swift
//  ZylixUITests
//
//  E2E Testing Context for Zylix-based iOS apps.
//  Provides unified testing helpers for XCUITest integration.
//

import XCTest

// MARK: - App State for Testing

/// Application state for E2E testing
public enum ZylixTestAppState: String {
    case idle
    case loading
    case ready
    case error
    case unknown
}

// MARK: - Test Configuration

/// Configuration for Zylix E2E tests
public struct ZylixTestConfig {
    /// Default timeout for waiting operations
    public var defaultTimeout: TimeInterval = 10.0

    /// Screenshot capture on failure
    public var captureScreenshotOnFailure: Bool = true

    /// Log level for test output
    public var logLevel: LogLevel = .info

    /// Reset app state before each test
    public var resetStateBeforeTest: Bool = true

    public enum LogLevel: Int {
        case none = 0
        case error = 1
        case warning = 2
        case info = 3
        case debug = 4
    }

    public init() {}
}

// MARK: - Zylix Test Context

/// Main testing context for Zylix E2E tests
public class ZylixTestContext {

    // MARK: - Properties

    /// The XCUIApplication under test
    public let app: XCUIApplication

    /// Test configuration
    public var config: ZylixTestConfig

    /// Current app state
    public private(set) var currentState: ZylixTestAppState = .unknown

    /// Screenshot storage
    private var screenshots: [XCUIScreenshot] = []

    // MARK: - Initialization

    /// Initialize with app bundle identifier
    public init(bundleIdentifier: String? = nil, config: ZylixTestConfig = ZylixTestConfig()) {
        if let bundleId = bundleIdentifier {
            self.app = XCUIApplication(bundleIdentifier: bundleId)
        } else {
            self.app = XCUIApplication()
        }
        self.config = config
    }

    /// Initialize with existing XCUIApplication
    public init(app: XCUIApplication, config: ZylixTestConfig = ZylixTestConfig()) {
        self.app = app
        self.config = config
    }

    // MARK: - App Lifecycle

    /// Launch the app with optional arguments
    public func launch(arguments: [String] = [], environment: [String: String] = [:]) {
        app.launchArguments = arguments
        app.launchEnvironment = environment

        if config.resetStateBeforeTest {
            app.launchArguments.append("--reset-state")
        }

        app.launch()
        log("App launched", level: .info)
    }

    /// Terminate the app
    public func terminate() {
        app.terminate()
        log("App terminated", level: .info)
    }

    /// Verify that Zylix core is initialized
    public func verifyInitialization(timeout: TimeInterval? = nil) -> Bool {
        let timeoutValue = timeout ?? config.defaultTimeout

        // Wait for app to be ready
        let readyPredicate = NSPredicate(format: "exists == true")
        let mainWindow = app.windows.firstMatch

        let expectation = XCTNSPredicateExpectation(predicate: readyPredicate, object: mainWindow)
        let result = XCTWaiter.wait(for: [expectation], timeout: timeoutValue)

        if result == .completed {
            currentState = .ready
            log("Zylix core initialized successfully", level: .info)
            return true
        } else {
            currentState = .error
            log("Zylix core initialization failed", level: .error)
            return false
        }
    }

    /// Get current app state
    public func getState() -> ZylixTestAppState {
        if app.state == .runningForeground {
            return currentState
        } else if app.state == .notRunning {
            return .idle
        } else {
            return .unknown
        }
    }

    // MARK: - Interaction Helpers

    /// Simulate tap at coordinates
    public func simulateTap(x: CGFloat, y: CGFloat) {
        let normalized = app.coordinate(withNormalizedOffset: CGVector(dx: x, dy: y))
        normalized.tap()
        log("Tap at (\(x), \(y))", level: .debug)
    }

    /// Simulate tap on element
    public func tap(_ element: XCUIElement, timeout: TimeInterval? = nil) {
        let timeoutValue = timeout ?? config.defaultTimeout

        if element.waitForExistence(timeout: timeoutValue) {
            element.tap()
            log("Tapped element: \(element.identifier)", level: .debug)
        } else {
            log("Element not found for tap: \(element.identifier)", level: .error)
        }
    }

    /// Simulate double tap on element
    public func doubleTap(_ element: XCUIElement, timeout: TimeInterval? = nil) {
        let timeoutValue = timeout ?? config.defaultTimeout

        if element.waitForExistence(timeout: timeoutValue) {
            element.doubleTap()
            log("Double tapped element: \(element.identifier)", level: .debug)
        }
    }

    /// Simulate long press on element
    public func longPress(_ element: XCUIElement, duration: TimeInterval = 1.0, timeout: TimeInterval? = nil) {
        let timeoutValue = timeout ?? config.defaultTimeout

        if element.waitForExistence(timeout: timeoutValue) {
            element.press(forDuration: duration)
            log("Long pressed element: \(element.identifier)", level: .debug)
        }
    }

    /// Simulate swipe gesture
    public func swipe(_ direction: SwipeDirection, on element: XCUIElement? = nil) {
        let target = element ?? app

        switch direction {
        case .up:
            target.swipeUp()
        case .down:
            target.swipeDown()
        case .left:
            target.swipeLeft()
        case .right:
            target.swipeRight()
        }

        log("Swiped \(direction)", level: .debug)
    }

    public enum SwipeDirection {
        case up, down, left, right
    }

    /// Type text into element
    public func typeText(_ text: String, into element: XCUIElement, timeout: TimeInterval? = nil) {
        let timeoutValue = timeout ?? config.defaultTimeout

        if element.waitForExistence(timeout: timeoutValue) {
            element.tap()
            element.typeText(text)
            log("Typed text into element: \(element.identifier)", level: .debug)
        }
    }

    /// Clear text in element
    public func clearText(in element: XCUIElement, timeout: TimeInterval? = nil) {
        let timeoutValue = timeout ?? config.defaultTimeout

        if element.waitForExistence(timeout: timeoutValue) {
            element.tap()

            if let stringValue = element.value as? String {
                let deleteString = String(repeating: XCUIKeyboardKey.delete.rawValue, count: stringValue.count)
                element.typeText(deleteString)
            }

            log("Cleared text in element: \(element.identifier)", level: .debug)
        }
    }

    // MARK: - Wait Helpers

    /// Wait for element to exist
    public func waitForElement(_ element: XCUIElement, timeout: TimeInterval? = nil) -> Bool {
        let timeoutValue = timeout ?? config.defaultTimeout
        return element.waitForExistence(timeout: timeoutValue)
    }

    /// Wait for element to disappear
    public func waitForElementToDisappear(_ element: XCUIElement, timeout: TimeInterval? = nil) -> Bool {
        let timeoutValue = timeout ?? config.defaultTimeout
        let predicate = NSPredicate(format: "exists == false")
        let expectation = XCTNSPredicateExpectation(predicate: predicate, object: element)
        let result = XCTWaiter.wait(for: [expectation], timeout: timeoutValue)
        return result == .completed
    }

    /// Wait for state change
    public func waitForStateChange(to targetState: ZylixTestAppState, timeout: TimeInterval? = nil) -> Bool {
        let timeoutValue = timeout ?? config.defaultTimeout
        let startTime = Date()

        while Date().timeIntervalSince(startTime) < timeoutValue {
            if getState() == targetState {
                log("State changed to \(targetState)", level: .debug)
                return true
            }
            Thread.sleep(forTimeInterval: 0.1)
        }

        log("Timeout waiting for state: \(targetState)", level: .warning)
        return false
    }

    /// Wait for element property
    public func waitForProperty<T: Equatable>(
        _ keyPath: KeyPath<XCUIElement, T>,
        of element: XCUIElement,
        toBe value: T,
        timeout: TimeInterval? = nil
    ) -> Bool {
        let timeoutValue = timeout ?? config.defaultTimeout
        let startTime = Date()

        while Date().timeIntervalSince(startTime) < timeoutValue {
            if element[keyPath: keyPath] == value {
                return true
            }
            Thread.sleep(forTimeInterval: 0.1)
        }

        return false
    }

    // MARK: - Element Query Helpers

    /// Find element by accessibility identifier
    public func element(withIdentifier id: String) -> XCUIElement {
        return app.descendants(matching: .any).matching(identifier: id).firstMatch
    }

    /// Find button by label
    public func button(withLabel label: String) -> XCUIElement {
        return app.buttons[label]
    }

    /// Find text field by identifier
    public func textField(withIdentifier id: String) -> XCUIElement {
        return app.textFields[id]
    }

    /// Find static text containing string
    public func staticText(containing text: String) -> XCUIElement {
        return app.staticTexts.containing(NSPredicate(format: "label CONTAINS[c] %@", text)).firstMatch
    }

    /// Find all elements of type
    public func elements(ofType type: XCUIElement.ElementType) -> XCUIElementQuery {
        return app.descendants(matching: type)
    }

    // MARK: - Assertion Helpers

    /// Assert element exists
    public func assertExists(_ element: XCUIElement, message: String = "", file: StaticString = #file, line: UInt = #line) {
        XCTAssertTrue(element.exists, message.isEmpty ? "Element should exist" : message, file: file, line: line)
    }

    /// Assert element does not exist
    public func assertNotExists(_ element: XCUIElement, message: String = "", file: StaticString = #file, line: UInt = #line) {
        XCTAssertFalse(element.exists, message.isEmpty ? "Element should not exist" : message, file: file, line: line)
    }

    /// Assert element is hittable
    public func assertHittable(_ element: XCUIElement, message: String = "", file: StaticString = #file, line: UInt = #line) {
        XCTAssertTrue(element.isHittable, message.isEmpty ? "Element should be hittable" : message, file: file, line: line)
    }

    /// Assert element is enabled
    public func assertEnabled(_ element: XCUIElement, message: String = "", file: StaticString = #file, line: UInt = #line) {
        XCTAssertTrue(element.isEnabled, message.isEmpty ? "Element should be enabled" : message, file: file, line: line)
    }

    /// Assert element label equals
    public func assertLabel(_ element: XCUIElement, equals expected: String, file: StaticString = #file, line: UInt = #line) {
        XCTAssertEqual(element.label, expected, "Element label should be '\(expected)'", file: file, line: line)
    }

    /// Assert element value equals
    public func assertValue(_ element: XCUIElement, equals expected: String, file: StaticString = #file, line: UInt = #line) {
        XCTAssertEqual(element.value as? String, expected, "Element value should be '\(expected)'", file: file, line: line)
    }

    // MARK: - Screenshot Helpers

    /// Capture screenshot
    public func captureScreenshot(name: String = "screenshot") -> XCUIScreenshot {
        let screenshot = app.screenshot()
        screenshots.append(screenshot)
        log("Screenshot captured: \(name)", level: .debug)
        return screenshot
    }

    /// Attach screenshot to test report
    public func attachScreenshot(_ screenshot: XCUIScreenshot, name: String = "screenshot") -> XCTAttachment {
        let attachment = XCTAttachment(screenshot: screenshot)
        attachment.name = name
        attachment.lifetime = .keepAlways
        return attachment
    }

    /// Capture and attach screenshot
    public func captureAndAttach(name: String = "screenshot") -> XCTAttachment {
        let screenshot = captureScreenshot(name: name)
        return attachScreenshot(screenshot, name: name)
    }

    // MARK: - Logging

    private func log(_ message: String, level: ZylixTestConfig.LogLevel) {
        guard level.rawValue <= config.logLevel.rawValue else { return }

        let prefix: String
        switch level {
        case .none: return
        case .error: prefix = "[ERROR]"
        case .warning: prefix = "[WARN]"
        case .info: prefix = "[INFO]"
        case .debug: prefix = "[DEBUG]"
        }

        print("\(prefix) ZylixTest: \(message)")
    }
}

// MARK: - Zylix Component Testing Helpers

extension ZylixTestContext {

    /// Verify Zylix component exists by type
    public func verifyComponent(type: String, identifier: String? = nil) -> Bool {
        let query: XCUIElementQuery

        if let id = identifier {
            query = app.descendants(matching: .any).matching(identifier: id)
        } else {
            query = app.descendants(matching: .any).matching(
                NSPredicate(format: "identifier CONTAINS[c] %@", "zylix-\(type)")
            )
        }

        return query.count > 0
    }

    /// Get Zylix component by type and identifier
    public func component(type: String, identifier: String) -> XCUIElement {
        return element(withIdentifier: "zylix-\(type)-\(identifier)")
    }

    /// Verify button component
    public func verifyButton(identifier: String, label: String? = nil) -> Bool {
        let button = component(type: "button", identifier: identifier)

        if !button.exists {
            return false
        }

        if let expectedLabel = label {
            return button.label == expectedLabel
        }

        return true
    }

    /// Verify text component
    public func verifyText(identifier: String, contains text: String) -> Bool {
        let textElement = component(type: "text", identifier: identifier)

        guard textElement.exists else { return false }

        return textElement.label.contains(text)
    }

    /// Verify input component
    public func verifyInput(identifier: String, value: String? = nil) -> Bool {
        let input = component(type: "input", identifier: identifier)

        guard input.exists else { return false }

        if let expectedValue = value {
            return (input.value as? String) == expectedValue
        }

        return true
    }
}

// MARK: - Test Case Base Class

/// Base class for Zylix XCUITest test cases
open class ZylixUITestCase: XCTestCase {

    /// Shared test context
    public var context: ZylixTestContext!

    /// Test configuration
    open var testConfig: ZylixTestConfig {
        return ZylixTestConfig()
    }

    /// App bundle identifier (override if needed)
    open var bundleIdentifier: String? {
        return nil
    }

    open override func setUpWithError() throws {
        try super.setUpWithError()

        continueAfterFailure = false

        context = ZylixTestContext(bundleIdentifier: bundleIdentifier, config: testConfig)
        context.launch()

        XCTAssertTrue(context.verifyInitialization(), "Zylix core should initialize")
    }

    open override func tearDownWithError() throws {
        if testConfig.captureScreenshotOnFailure {
            let attachment = context.captureAndAttach(name: "final-state")
            add(attachment)
        }

        context.terminate()
        context = nil

        try super.tearDownWithError()
    }
}

File: platforms/android/zylix-android/src/androidTest/kotlin/com/zylix/test/ZylixTestContext.kt
package com.zylix.test

import android.app.Activity
import android.content.Intent
import android.view.View
import androidx.test.core.app.ActivityScenario
import androidx.test.core.app.ApplicationProvider
import androidx.test.espresso.Espresso
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.UiController
import androidx.test.espresso.ViewAction
import androidx.test.espresso.ViewInteraction
import androidx.test.espresso.action.ViewActions
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.platform.app.InstrumentationRegistry
import androidx.test.uiautomator.UiDevice
import org.hamcrest.Matcher
import org.hamcrest.Matchers.*
import java.io.File
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * Application state for E2E testing
 */
enum class ZylixTestAppState {
    IDLE,
    LOADING,
    READY,
    ERROR,
    UNKNOWN
}

/**
 * Configuration for Zylix E2E tests
 */
data class ZylixTestConfig(
    /** Default timeout for waiting operations (ms) */
    val defaultTimeout: Long = 10_000L,

    /** Screenshot capture on failure */
    val captureScreenshotOnFailure: Boolean = true,

    /** Log level for test output */
    val logLevel: LogLevel = LogLevel.INFO,

    /** Reset app state before each test */
    val resetStateBeforeTest: Boolean = true
) {
    enum class LogLevel(val level: Int) {
        NONE(0),
        ERROR(1),
        WARNING(2),
        INFO(3),
        DEBUG(4)
    }
}

/**
 * Swipe direction for gestures
 */
enum class SwipeDirection {
    UP, DOWN, LEFT, RIGHT
}

/**
 * Main testing context for Zylix E2E tests on Android.
 *
 * Provides unified testing helpers for Espresso integration.
 *
 * @param activityClass The activity class to launch for testing
 * @param config Test configuration options
 */
class ZylixTestContext<T : Activity>(
    private val activityClass: Class<T>,
    val config: ZylixTestConfig = ZylixTestConfig()
) {
    private var scenario: ActivityScenario<T>? = null
    private var currentState: ZylixTestAppState = ZylixTestAppState.UNKNOWN
    private val screenshots: MutableList<File> = mutableListOf()
    private val uiDevice: UiDevice by lazy {
        UiDevice.getInstance(InstrumentationRegistry.getInstrumentation())
    }

    // MARK: - App Lifecycle

    /**
     * Launch the activity with optional intent extras
     */
    fun launch(
        intentExtras: Map<String, Any> = emptyMap(),
        launchFlags: Int? = null
    ) {
        val intent = Intent(ApplicationProvider.getApplicationContext(), activityClass).apply {
            intentExtras.forEach { (key, value) ->
                when (value) {
                    is String -> putExtra(key, value)
                    is Int -> putExtra(key, value)
                    is Long -> putExtra(key, value)
                    is Boolean -> putExtra(key, value)
                    is Float -> putExtra(key, value)
                    is Double -> putExtra(key, value)
                }
            }
            launchFlags?.let { flags = it }

            if (config.resetStateBeforeTest) {
                putExtra("--reset-state", true)
            }
        }

        scenario = ActivityScenario.launch(intent)
        log("Activity launched", ZylixTestConfig.LogLevel.INFO)
    }

    /**
     * Close the activity
     */
    fun close() {
        scenario?.close()
        scenario = null
        log("Activity closed", ZylixTestConfig.LogLevel.INFO)
    }

    /**
     * Verify that Zylix core is initialized
     */
    fun verifyInitialization(timeout: Long? = null): Boolean {
        val timeoutValue = timeout ?: config.defaultTimeout
        val startTime = System.currentTimeMillis()

        while (System.currentTimeMillis() - startTime < timeoutValue) {
            try {
                // Wait for activity to be in resumed state
                scenario?.onActivity { activity ->
                    if (activity.window.decorView.isShown) {
                        currentState = ZylixTestAppState.READY
                    }
                }

                if (currentState == ZylixTestAppState.READY) {
                    log("Zylix core initialized successfully", ZylixTestConfig.LogLevel.INFO)
                    return true
                }
            } catch (e: Exception) {
                log("Initialization check error: ${e.message}", ZylixTestConfig.LogLevel.DEBUG)
            }

            Thread.sleep(100)
        }

        currentState = ZylixTestAppState.ERROR
        log("Zylix core initialization failed", ZylixTestConfig.LogLevel.ERROR)
        return false
    }

    /**
     * Get current app state
     */
    fun getState(): ZylixTestAppState = currentState

    // MARK: - Interaction Helpers

    /**
     * Simulate tap at coordinates
     */
    fun simulateTap(x: Float, y: Float) {
        uiDevice.click(x.toInt(), y.toInt())
        log("Tap at ($x, $y)", ZylixTestConfig.LogLevel.DEBUG)
    }

    /**
     * Tap on view matching the given matcher
     */
    fun tap(viewMatcher: Matcher<View>, timeout: Long? = null): ViewInteraction {
        waitForView(viewMatcher, timeout)
        return onView(viewMatcher).perform(ViewActions.click()).also {
            log("Tapped view", ZylixTestConfig.LogLevel.DEBUG)
        }
    }

    /**
     * Tap on view with specific ID
     */
    fun tapById(viewId: Int, timeout: Long? = null): ViewInteraction {
        return tap(withId(viewId), timeout)
    }

    /**
     * Tap on view with specific text
     */
    fun tapByText(text: String, timeout: Long? = null): ViewInteraction {
        return tap(withText(text), timeout)
    }

    /**
     * Tap on view with content description
     */
    fun tapByContentDescription(description: String, timeout: Long? = null): ViewInteraction {
        return tap(withContentDescription(description), timeout)
    }

    /**
     * Double tap on view
     */
    fun doubleTap(viewMatcher: Matcher<View>, timeout: Long? = null): ViewInteraction {
        waitForView(viewMatcher, timeout)
        return onView(viewMatcher).perform(ViewActions.doubleClick()).also {
            log("Double tapped view", ZylixTestConfig.LogLevel.DEBUG)
        }
    }

    /**
     * Long press on view
     */
    fun longPress(viewMatcher: Matcher<View>, timeout: Long? = null): ViewInteraction {
        waitForView(viewMatcher, timeout)
        return onView(viewMatcher).perform(ViewActions.longClick()).also {
            log("Long pressed view", ZylixTestConfig.LogLevel.DEBUG)
        }
    }

    /**
     * Swipe gesture on view
     */
    fun swipe(direction: SwipeDirection, viewMatcher: Matcher<View>? = null): ViewInteraction {
        val target = viewMatcher ?: isRoot()
        val action = when (direction) {
            SwipeDirection.UP -> ViewActions.swipeUp()
            SwipeDirection.DOWN -> ViewActions.swipeDown()
            SwipeDirection.LEFT -> ViewActions.swipeLeft()
            SwipeDirection.RIGHT -> ViewActions.swipeRight()
        }
        return onView(target).perform(action).also {
            log("Swiped $direction", ZylixTestConfig.LogLevel.DEBUG)
        }
    }

    /**
     * Type text into view
     */
    fun typeText(text: String, viewMatcher: Matcher<View>, timeout: Long? = null): ViewInteraction {
        waitForView(viewMatcher, timeout)
        return onView(viewMatcher)
            .perform(ViewActions.click())
            .perform(ViewActions.typeText(text))
            .also {
                log("Typed text into view", ZylixTestConfig.LogLevel.DEBUG)
            }
    }

    /**
     * Type text into view by ID
     */
    fun typeTextById(text: String, viewId: Int, timeout: Long? = null): ViewInteraction {
        return typeText(text, withId(viewId), timeout)
    }

    /**
     * Clear text in view
     */
    fun clearText(viewMatcher: Matcher<View>, timeout: Long? = null): ViewInteraction {
        waitForView(viewMatcher, timeout)
        return onView(viewMatcher)
            .perform(ViewActions.click())
            .perform(ViewActions.clearText())
            .also {
                log("Cleared text in view", ZylixTestConfig.LogLevel.DEBUG)
            }
    }

    /**
     * Replace text in view
     */
    fun replaceText(text: String, viewMatcher: Matcher<View>, timeout: Long? = null): ViewInteraction {
        waitForView(viewMatcher, timeout)
        return onView(viewMatcher)
            .perform(ViewActions.replaceText(text))
            .also {
                log("Replaced text in view", ZylixTestConfig.LogLevel.DEBUG)
            }
    }

    /**
     * Close soft keyboard
     */
    fun closeKeyboard() {
        Espresso.closeSoftKeyboard()
        log("Closed keyboard", ZylixTestConfig.LogLevel.DEBUG)
    }

    /**
     * Press back button
     */
    fun pressBack() {
        Espresso.pressBack()
        log("Pressed back", ZylixTestConfig.LogLevel.DEBUG)
    }

    // MARK: - Wait Helpers

    /**
     * Wait for view to be visible
     */
    fun waitForView(
        viewMatcher: Matcher<View>,
        timeout: Long? = null
    ): Boolean {
        val timeoutValue = timeout ?: config.defaultTimeout
        val startTime = System.currentTimeMillis()

        while (System.currentTimeMillis() - startTime < timeoutValue) {
            try {
                onView(viewMatcher).check(matches(isDisplayed()))
                return true
            } catch (e: Exception) {
                Thread.sleep(100)
            }
        }

        log("Timeout waiting for view", ZylixTestConfig.LogLevel.WARNING)
        return false
    }

    /**
     * Wait for view to disappear
     */
    fun waitForViewToDisappear(
        viewMatcher: Matcher<View>,
        timeout: Long? = null
    ): Boolean {
        val timeoutValue = timeout ?: config.defaultTimeout
        val startTime = System.currentTimeMillis()

        while (System.currentTimeMillis() - startTime < timeoutValue) {
            try {
                onView(viewMatcher).check(matches(not(isDisplayed())))
                return true
            } catch (e: Exception) {
                // View is still visible
                try {
                    onView(viewMatcher).check(matches(isDisplayed()))
                    Thread.sleep(100)
                } catch (e2: Exception) {
                    // View doesn't exist, which means it's disappeared
                    return true
                }
            }
        }

        log("Timeout waiting for view to disappear", ZylixTestConfig.LogLevel.WARNING)
        return false
    }

    /**
     * Wait for state change
     */
    fun waitForStateChange(
        targetState: ZylixTestAppState,
        timeout: Long? = null
    ): Boolean {
        val timeoutValue = timeout ?: config.defaultTimeout
        val startTime = System.currentTimeMillis()

        while (System.currentTimeMillis() - startTime < timeoutValue) {
            if (getState() == targetState) {
                log("State changed to $targetState", ZylixTestConfig.LogLevel.DEBUG)
                return true
            }
            Thread.sleep(100)
        }

        log("Timeout waiting for state: $targetState", ZylixTestConfig.LogLevel.WARNING)
        return false
    }

    /**
     * Wait for specific duration
     */
    fun waitFor(milliseconds: Long) {
        Thread.sleep(milliseconds)
    }

    /**
     * Idle sync - wait for all pending tasks to complete
     */
    fun idleSync() {
        InstrumentationRegistry.getInstrumentation().waitForIdleSync()
    }

    // MARK: - View Query Helpers

    /**
     * Check if view with ID exists
     */
    fun viewExists(viewId: Int): Boolean {
        return try {
            onView(withId(viewId)).check(matches(isDisplayed()))
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Check if view with text exists
     */
    fun viewWithTextExists(text: String): Boolean {
        return try {
            onView(withText(text)).check(matches(isDisplayed()))
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Get view interaction by ID
     */
    fun view(viewId: Int): ViewInteraction = onView(withId(viewId))

    /**
     * Get view interaction by text
     */
    fun viewWithText(text: String): ViewInteraction = onView(withText(text))

    /**
     * Get view interaction by content description
     */
    fun viewWithDescription(description: String): ViewInteraction =
        onView(withContentDescription(description))

    /**
     * Get view with tag
     */
    fun viewWithTag(tag: Any): ViewInteraction = onView(withTagValue(`is`(tag)))

    // MARK: - Assertion Helpers

    /**
     * Assert view is displayed
     */
    fun assertDisplayed(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(isDisplayed()))
    }

    /**
     * Assert view with ID is displayed
     */
    fun assertDisplayedById(viewId: Int): ViewInteraction {
        return assertDisplayed(withId(viewId))
    }

    /**
     * Assert view with text is displayed
     */
    fun assertDisplayedByText(text: String): ViewInteraction {
        return assertDisplayed(withText(text))
    }

    /**
     * Assert view is not displayed
     */
    fun assertNotDisplayed(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(not(isDisplayed())))
    }

    /**
     * Assert view is enabled
     */
    fun assertEnabled(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(isEnabled()))
    }

    /**
     * Assert view is not enabled
     */
    fun assertNotEnabled(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(not(isEnabled())))
    }

    /**
     * Assert view has text
     */
    fun assertHasText(viewMatcher: Matcher<View>, expectedText: String): ViewInteraction {
        return onView(viewMatcher).check(matches(withText(expectedText)))
    }

    /**
     * Assert view contains text
     */
    fun assertContainsText(viewMatcher: Matcher<View>, text: String): ViewInteraction {
        return onView(viewMatcher).check(matches(withText(containsString(text))))
    }

    /**
     * Assert view is clickable
     */
    fun assertClickable(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(isClickable()))
    }

    /**
     * Assert view is checked (for checkboxes, radio buttons)
     */
    fun assertChecked(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(isChecked()))
    }

    /**
     * Assert view is not checked
     */
    fun assertNotChecked(viewMatcher: Matcher<View>): ViewInteraction {
        return onView(viewMatcher).check(matches(not(isChecked())))
    }

    // MARK: - Screenshot Helpers

    /**
     * Capture screenshot
     */
    fun captureScreenshot(name: String = "screenshot"): File? {
        return try {
            val timestamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())
            val fileName = "${name}_$timestamp.png"
            val screenshotDir = File(
                InstrumentationRegistry.getInstrumentation().targetContext.filesDir,
                "screenshots"
            )
            screenshotDir.mkdirs()

            val screenshotFile = File(screenshotDir, fileName)
            uiDevice.takeScreenshot(screenshotFile)
            screenshots.add(screenshotFile)

            log("Screenshot captured: $fileName", ZylixTestConfig.LogLevel.DEBUG)
            screenshotFile
        } catch (e: Exception) {
            log("Failed to capture screenshot: ${e.message}", ZylixTestConfig.LogLevel.ERROR)
            null
        }
    }

    /**
     * Get all captured screenshots
     */
    fun getScreenshots(): List<File> = screenshots.toList()

    // MARK: - Zylix Component Helpers

    /**
     * Verify Zylix component exists by tag pattern
     */
    fun verifyComponent(type: String, identifier: String? = null): Boolean {
        val tag = if (identifier != null) {
            "zylix-$type-$identifier"
        } else {
            "zylix-$type"
        }

        return try {
            onView(withTagValue(containsString(tag))).check(matches(isDisplayed()))
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Get Zylix component by type and identifier
     */
    fun component(type: String, identifier: String): ViewInteraction {
        return viewWithTag("zylix-$type-$identifier")
    }

    /**
     * Verify button component
     */
    fun verifyButton(identifier: String, text: String? = null): Boolean {
        return try {
            val button = component("button", identifier)
            button.check(matches(isDisplayed()))
            text?.let { button.check(matches(withText(it))) }
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Verify text component
     */
    fun verifyText(identifier: String, contains: String): Boolean {
        return try {
            component("text", identifier)
                .check(matches(withText(containsString(contains))))
            true
        } catch (e: Exception) {
            false
        }
    }

    /**
     * Verify input component
     */
    fun verifyInput(identifier: String, value: String? = null): Boolean {
        return try {
            val input = component("input", identifier)
            input.check(matches(isDisplayed()))
            value?.let { input.check(matches(withText(it))) }
            true
        } catch (e: Exception) {
            false
        }
    }

    // MARK: - Logging

    private fun log(message: String, level: ZylixTestConfig.LogLevel) {
        if (level.level <= config.logLevel.level) {
            val prefix = when (level) {
                ZylixTestConfig.LogLevel.NONE -> return
                ZylixTestConfig.LogLevel.ERROR -> "[ERROR]"
                ZylixTestConfig.LogLevel.WARNING -> "[WARN]"
                ZylixTestConfig.LogLevel.INFO -> "[INFO]"
                ZylixTestConfig.LogLevel.DEBUG -> "[DEBUG]"
            }
            println("$prefix ZylixTest: $message")
        }
    }

    // MARK: - Activity Access

    /**
     * Execute action on activity
     */
    fun <R> onActivity(action: (T) -> R): R? {
        var result: R? = null
        scenario?.onActivity { activity ->
            result = action(activity)
        }
        return result
    }
}

/**
 * Custom ViewAction for waiting
 */
fun waitFor(delay: Long): ViewAction {
    return object : ViewAction {
        override fun getConstraints(): Matcher<View> = isRoot()

        override fun getDescription(): String = "Wait for $delay milliseconds"

        override fun perform(uiController: UiController, view: View?) {
            uiController.loopMainThreadForAtLeast(delay)
        }
    }
}
